File: 0752-open-the-lock/0752-open-the-lock.java
class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> deadSet = new HashSet<>(Arrays.asList(deadends));
        if(deadSet.contains("0000")) return -1;
        Queue<String> q = new LinkedList<>(Collections.singletonList("0000"));
        for(int steps = 0; !q.isEmpty(); steps++) {
            for(int size = q.size(); size > 0; --size) {
                String curr = q.poll();
                if(curr.equals(target)) return steps;
                for(String nei : neighbors(curr)) {
                    if(deadSet.contains(nei)) continue;
                    deadSet.add(nei);
                    q.offer(nei);
                }
            }
        }
        return -1;
    }
    
    List<String> neighbors(String code) {
        List<String> results = new ArrayList<>();
        for(int i = 0; i < 4; ++i)
            for(int diff = -1; diff <= 1; diff += 2) {
                results.add(code.substring(0,i) + (code.charAt(i) - '0' + diff + 10)%10 + "" + code.substring(i+1));
            }
        return results;
    }
}

File: 0752-open-the-lock/0752-open-the-lock.py
class Solution(object):
    def openLock(self, deadends, target):
        """
        :type deadends: List[str]
        :type target: str
        :rtype: int
        """
        
        deadSet = set(deadends)
        if "0000" in deadSet: return -1
        q = deque(["0000"])
        steps = 0
        while q:
            for _ in range(len(q)):
                curr = q.popleft()
                if curr == target: return steps
                for nei in self.neighbour(curr):
                    if nei in deadSet: continue
                    deadSet.add(nei)
                    q.append(nei)
            steps += 1 
        return -1
    def neighbour(self, code):
        for i in range(4):
            x = int(code[i])
            for diff in [-1,1]:
                y = (x + diff + 10)%10
                yield code[:i]+str(y)+code[i+1:]